### PR TITLE
Support `@readonly` phpdoc

### PR DESCRIPTION
**Pull request by herndlm**
Same as psalm does it, see https://psalm.dev/docs/annotating_code/supported_annotations/#psalm-readonly-and-readonly

I guess this would be very useful for pre PHP 8.1 and/or to safely enable native `readonly`. But I think this is also the first step in the direction of a `@phpstan-immutable` phpdoc on class level, but that's a different thing.

~~If I understood things right this should not interfere with some of the other rules that are related to native `readonly` because the phpdoc is not used in https://github.com/phpstan/phpstan-src/blob/1.6.7/src/Node/ClassPropertyNode.php#L77. I added tests to check that where it made sense to me.~~ irrelevant, because the phpdoc rules were separated later on :)

Refs: https://github.com/phpstan/phpstan/issues/4082

### CODE REVIEW

**Comment by staabm** [comment]
I had a discussion in the past about implementing it.

These are the main directions, which ondrey gave me:

> My thinking is that there are separate rules: for uninitialized properties that are opt-in with a config option
for native readonly properties checked on PHP 8.1+
and a different (but similar) rule would have to be added for 
@readonly PHPDoc properties.
but it's less relevant over time as people upgrade their PHP version and can start using the native feature. 
if you look at UninitializedPropertyRule + ReadOnlyPropertyAssignRule + MissingReadOnlyPropertyAssignRule you can see the differences and the similarilties

I don't know whether this information is still valid as is... just to let you know.

**Comment by ondrejmirtes** [comment]
I think for this to work, we need a new method `PhpPropertyReflection::isReadOnlyByPhpDoc()` and `PHPStan\Node\ClassPropertyNode::isReadOnlyByPhpDoc()` and a new set of rules (or modified implementation of the current ones with a new set of tests):

* UninitializedPropertyRule
* MissingReadOnlyPropertyAssignRule
* ReadOnlyPropertyAssignRefRule
* ReadOnlyPropertyAssignRule

I really want to separate the rules for "this is native readonly keyword" and "this is done by PHPDoc" as they can diverge over time. I don't want the current `isReadOnly()` methods to also report `true` for PHPDocs.

File: src/Node/ClassPropertiesNode.php [hunk 1]
```diff
@@ -96,7 +96,7 @@ public function getUninitializedProperties(
 			if ($property->isStatic()) {
 				continue;
 			}
-			if ($property->getNativeType() === null) {
+			if ($property->getNativeType() === null && !$property->isReadonlyByPhpDoc()) {
 				continue;
 			}
```
Comment (herndlm) [inline src/Node/ClassPropertiesNode.php]: this is weird, isn't it? it is currently needed to make 2 out of 3 cases (uninitialized and premature-accessed properties)  of `MissingReadOnlyPropertyAssignRule` work. But this is very native readonly specific IMO and could be annoying, maybe we should not support phpdoc readonly there? 
UPDATE: I was mistaken, this is also indirectly used for the 3rd case, meaning without it `MissingReadOnlyPropertyAssignRule` does not work at all :/ But maybe this is fine? the whole rule is basically made for the native assignment specific behaviour.

**Comment by herndlm** [comment]
I separated native and phpdoc readonly, as requested, and for simplicity I chose to adapt the current existing rules. I also added new test sets that are basically duplicates of the current ones minus PHP 7.4+ features (native property types, native readonly, enums).

Good idea with the separation for sure, see https://github.com/phpstan/phpstan-src/pull/1295#discussion_r873073588 where I decided to not add phpdoc readonly support for `MissingReadOnlyPropertyAssignRule` and `UninitializedPropertyRule` because things start to get mixed up with native readonly. I could add support back without that weird change, it would only work combined with native type hints if I'm not mistaken. Maybe that's fine after all. I can't decide :)

What do you think, is this going in the right direction?

File: src/Node/ClassPropertiesNode.php (hunk 1 above)
Comment (herndlm) [reply src/Node/ClassPropertiesNode.php]: I decided to drop phpdoc readonly support in `MissingReadOnlyPropertyAssignRule` again since it seems to handle specific native readonly behaviour. UPDATE: this affected `UninitializedPropertyRule` as well.

**Comment by herndlm** [comment]
Sorry for my impatience, I hate that xD
But since I see the 1.7 release nicely finishing up - do you think this can still be squeezed in? I'll prioritize any requested changes

**Comment by ondrejmirtes** [comment]
> do you think this can still be squeezed in? I'll prioritize any requested changes

Let me assure you that you can keep calm because it mostly doesn't matter :)

* No one says that PHPStan 1.8.0 cannot be released like 5 days after 1.7.0 :)
* No one says that this PR cannot be released in 1.7.1 for example :)

File: src/Analyser/NodeScopeResolver.php [hunk 2]
```diff
@@ -3839,9 +3838,9 @@ private function processNodesForTraitUse($node, ClassReflection $traitReflection
 	}
 
 	/**
-	 * @return array{TemplateTypeMap, Type[], ?Type, ?Type, ?string, bool, bool, bool, bool|null}
+	 * @return array{TemplateTypeMap, Type[], ?Type, ?Type, ?string, bool, bool, bool, bool|null, bool, string|null}
 	 */
-	public function getPhpDocs(Scope $scope, Node\FunctionLike $functionLike): array
+	public function getPhpDocs(Scope $scope, Node $node): array
```
Comment (ondrejmirtes) [inline src/Analyser/NodeScopeResolver.php]: Can we make this FunctionLike/Property union type? It can be native, it's gonna be downgraded automatically.

File: src/Rules/Properties/ReadOnlyPropertyAssignRefRule.php [hunk 3]
```diff
@@ -41,7 +41,7 @@ public function processNode(Node $node, Scope $scope): array
 			if (!$scope->canAccessProperty($propertyReflection)) {
 				continue;
 			}
-			if (!$nativeReflection->isReadOnly()) {
+			if (!$nativeReflection->isReadOnly() && !$nativeReflection->isReadOnlyByPhpDoc()) {
```
Comment (ondrejmirtes) [inline src/Rules/Properties/ReadOnlyPropertyAssignRefRule.php]: I really wanted a separate set of rules for this. So please copy this rule, it can be named something like `ReadOnlyByPhpDocProperty...` and write a separate testcase for it. And register it only for bleeding edge :)

File: src/Rules/Properties/ReadOnlyPropertyAssignRule.php [hunk 4]
```diff
@@ -45,7 +45,7 @@ public function processNode(Node $node, Scope $scope): array
 			if (!$scope->canAccessProperty($propertyReflection)) {
 				continue;
 			}
-			if (!$nativeReflection->isReadOnly()) {
+			if (!$nativeReflection->isReadOnly() && !$nativeReflection->isReadOnlyByPhpDoc()) {
```
Comment (ondrejmirtes) [inline src/Rules/Properties/ReadOnlyPropertyAssignRule.php]: dtto

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
> where I decided to not add phpdoc readonly support for MissingReadOnlyPropertyAssignRule and UninitializedPropertyRule

With my proposal of separate rules, this should probably be done, right?

File: src/Rules/Properties/ReadOnlyPropertyAssignRefRule.php (hunk 3 above)
Comment (herndlm) [reply src/Rules/Properties/ReadOnlyPropertyAssignRefRule.php]: Of course, makes sense. I was not fully sure what you'd prefer.
Regarding bleeding edge I hoped this could be avoided in a minor release. Is the reasoning mostly that already used readonly docs could trigger new annoying errors?

Comment (ondrejmirtes) [reply src/Rules/Properties/ReadOnlyPropertyAssignRefRule.php]: It can definitely be annoying and show unexpected errors. For users that use both Psalm and PHPStan, but:

1) Have Psalm errors ignored by the baseline, PHPStan would start reporting new errors about readonly
2) There can be implementation differences between Psalm and PHPStan, so something that's okay for Psalm might be reported by PHPStan

With bleedingEdge we're gonna gather useful feedback by people that willingly early-adopt this stuff 😊

File: conf/config.level0.neon [hunk 5]
```diff
@@ -174,6 +176,13 @@ services:
 		tags:
 			- phpstan.rules.rule
 
+	-
+		class: PHPStan\Rules\Properties\MissingReadOnlyByPhpDocPropertyAssignRule
+		arguments:
+			additionalConstructors: %additionalConstructors%
+		tags:
+			- phpstan.rules.rule
```
Comment (ondrejmirtes) [inline conf/config.level0.neon]: The `tags` section shouldn't be here - it should only be enabled in bleedingEdge which is already handled by 
`conditionalTags`.

File: src/Node/ClassPropertyNode.php [hunk 6]
```diff
@@ -21,6 +21,7 @@ public function __construct(
 		private ?string $phpDoc,
 		private bool $isPromoted,
 		Node $originalNode,
+		private bool $isReadonlyByPhpDoc = false,
```
Comment (ondrejmirtes) [inline src/Node/ClassPropertyNode.php]: There isn't a reason for the parameter to be optional

File: src/PhpDoc/PhpDocNodeResolver.php [hunk 7]
```diff
@@ -452,6 +452,17 @@ public function resolveIsImpure(PhpDocNode $phpDocNode): bool
 		return false;
 	}
 
+	public function resolveIsReadOnly(PhpDocNode $phpDocNode): bool
+	{
+		foreach ($phpDocNode->getTags() as $phpDocTagNode) {
+			if (in_array($phpDocTagNode->name, ['@readonly', '@psalm-readonly', '@phpstan-readonly'], true)) {
```
Comment (ondrejmirtes) [inline src/PhpDoc/PhpDocNodeResolver.php]: Would be nicer to match `resolveHasConsistentConstructor` below here - `getTagsByName` + `count`.

File: src/Rules/Properties/MissingReadOnlyByPhpDocPropertyAssignRule.php [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+		$classReflection = $scope->getClassReflection();
+		[$properties, $prematureAccess, $additionalAssigns] = $node->getUninitializedProperties($scope, $this->getConstructors($classReflection), []);
+
+		$errors = [];
+		foreach ($properties as $propertyName => $propertyNode) {
+			if (!$propertyNode->isReadOnlyByPhpDoc()) {
+				continue;
+			}
+			$errors[] = RuleErrorBuilder::message(sprintf(
+				'Class %s has an uninitialized readonly property $%s. Assign it in the constructor.',
```
Comment (ondrejmirtes) [inline src/Rules/Properties/MissingReadOnlyByPhpDocPropertyAssignRule.php]: `readonly` -> `@readonly` - so that we have a slight difference and know it's from a PHPDoc :)

File: src/Rules/Properties/MissingReadOnlyByPhpDocPropertyAssignRule.php [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+				$propertyName,
+			))->line($propertyNode->getLine())->build();
+		}
+
+		foreach ($prematureAccess as [$propertyName, $line, $propertyNode]) {
+			if (!$propertyNode->isReadOnlyByPhpDoc()) {
+				continue;
+			}
+			$errors[] = RuleErrorBuilder::message(sprintf(
+				'Access to an uninitialized readonly property %s::$%s.',
```
Comment (ondrejmirtes) [inline src/Rules/Properties/MissingReadOnlyByPhpDocPropertyAssignRule.php]: `@readonly`

File: src/Rules/Properties/MissingReadOnlyByPhpDocPropertyAssignRule.php [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+				$propertyName,
+			))->line($line)->build();
+		}
+
+		foreach ($additionalAssigns as [$propertyName, $line, $propertyNode]) {
+			if (!$propertyNode->isReadOnlyByPhpDoc()) {
+				continue;
+			}
+			$errors[] = RuleErrorBuilder::message(sprintf(
+				'Readonly property %s::$%s is already assigned.',
```
Comment (ondrejmirtes) [inline src/Rules/Properties/MissingReadOnlyByPhpDocPropertyAssignRule.php]: `@readonly`

File: src/Rules/Properties/ReadOnlyByPhpDocPropertyAssignRefRule.php [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+			}
+			if (!$scope->canAccessProperty($propertyReflection)) {
+				continue;
+			}
+			if (!$nativeReflection->isReadOnlyByPhpDoc()) {
+				continue;
+			}
+
+			$declaringClass = $nativeReflection->getDeclaringClass();
+			$errors[] = RuleErrorBuilder::message(sprintf('Readonly property %s::$%s is assigned by reference.', $declaringClass->getDisplayName(), $propertyReflection->getName()))->build();
```
Comment (ondrejmirtes) [inline src/Rules/Properties/ReadOnlyByPhpDocPropertyAssignRefRule.php]: `@readonly`

File: src/Rules/Properties/UninitializedPropertyRule.php [hunk 12]
```diff
@@ -59,7 +59,7 @@ public function processNode(Node $node, Scope $scope): array
 		}
 
 		foreach ($prematureAccess as [$propertyName, $line, $propertyNode]) {
-			if ($propertyNode->isReadOnly()) {
+			if ($propertyNode->isReadOnly() || $propertyNode->isReadOnlyByPhpDoc()) {
```
Comment (herndlm) [inline src/Rules/Properties/UninitializedPropertyRule.php:62]: I kept this in / got it back on purpose. I think a separate rule does not make much sense, does it? This rule does not deal with `readonly` in particular, but it ignores it because other rules exist. I think the right thing to do, since now other readonly phpdoc rules exist, is to ignore this then here as well. Does that make sense?

File: src/Rules/Properties/ReadOnlyByPhpDocPropertyAssignRule.php [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+				continue;
+			}
+			if (!$nativeReflection->isReadOnlyByPhpDoc()) {
+				continue;
+			}
+
+			$declaringClass = $nativeReflection->getDeclaringClass();
+
+			if (!$scope->isInClass()) {
+				$errors[] = RuleErrorBuilder::message(sprintf('Readonly property %s::$%s is assigned outside of its declaring class.', $declaringClass->getDisplayName(), $propertyReflection->getName()))->build();
```
Comment (ondrejmirtes) [inline src/Rules/Properties/ReadOnlyByPhpDocPropertyAssignRule.php]: `@readonly`

File: src/Rules/Properties/ReadOnlyByPhpDocPropertyAssignRule.php [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+				continue;
+			}
+
+			$scopeClassReflection = $scope->getClassReflection();
+			if ($scopeClassReflection->getName() !== $declaringClass->getName()) {
+				$errors[] = RuleErrorBuilder::message(sprintf('Readonly property %s::$%s is assigned outside of its declaring class.', $declaringClass->getDisplayName(), $propertyReflection->getName()))->build();
```
Comment (ondrejmirtes) [inline src/Rules/Properties/ReadOnlyByPhpDocPropertyAssignRule.php]: `@readonly`

File: src/Rules/Properties/ReadOnlyByPhpDocPropertyAssignRule.php [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+			}
+
+			$scopeMethod = $scope->getFunction();
+			if (!$scopeMethod instanceof MethodReflection) {
+				throw new ShouldNotHappenException();
+			}
+
+			if (strtolower($scopeMethod->getName()) === '__construct' || strtolower($scopeMethod->getName()) === '__unserialize') {
+				if (!$scope->getType($propertyFetch->var) instanceof ThisType) {
+					$errors[] = RuleErrorBuilder::message(sprintf('Readonly property %s::$%s is not assigned on $this.', $declaringClass->getDisplayName(), $propertyReflection->getName()))->build();
```
Comment (ondrejmirtes) [inline src/Rules/Properties/ReadOnlyByPhpDocPropertyAssignRule.php]: `@readonly`

File: src/Rules/Properties/ReadOnlyByPhpDocPropertyAssignRule.php [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+				}
+
+				continue;
+			}
+
+			$errors[] = RuleErrorBuilder::message(sprintf('Readonly property %s::$%s is assigned outside of the constructor.', $declaringClass->getDisplayName(), $propertyReflection->getName()))->build();
```
Comment (ondrejmirtes) [inline src/Rules/Properties/ReadOnlyByPhpDocPropertyAssignRule.php]: `@readonly`

File: src/Rules/Properties/MissingReadOnlyByPhpDocPropertyAssignRule.php [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+			))->line($line)->build();
+		}
+
+		return $errors;
+	}
+
+	/**
+	 * @return string[]
+	 */
+	private function getConstructors(ClassReflection $classReflection): array
```
Comment (ondrejmirtes) [inline src/Rules/Properties/MissingReadOnlyByPhpDocPropertyAssignRule.php:88]: As a follow-up PR, we could clean up and DRY all the places with this code:

* MissingReadOnlyPropertyAssignRule
* MissingReadOnlyByPhpDocPropertyAssignRule
* UninitializedPropertyRule

Additionally, there's https://github.com/phpstan/phpstan/issues/6612 to take care of.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
Really nice! Almost there :)

File: src/Rules/Properties/UninitializedPropertyRule.php (hunk 12 above)
Comment (ondrejmirtes) [reply src/Rules/Properties/UninitializedPropertyRule.php:62]: Yeah, of course, this is about NOT reporting something so this change makes sense.

File: src/Node/ClassPropertyNode.php (hunk 6 above)
Comment (herndlm) [reply src/Node/ClassPropertyNode.php]: Oh, I would prefer to remove that default value. Is it fine even if we have `/** @api */`?

Comment (ondrejmirtes) [reply src/Node/ClassPropertyNode.php]: Constructors are special :) https://phpstan.org/developing-extensions/backward-compatibility-promise

> Objects of classes with `@api` in their PHPDoc cannot be created via `new` unless there's also `@api` above the public constructor. This encourages using [dependency injection](/developing-extensions/dependency-injection-configuration) in case the class is registered as a service in the dependency injection container.

File: src/Rules/Properties/MissingReadOnlyByPhpDocPropertyAssignRule.php (hunk 17 above)
Comment (herndlm) [reply src/Rules/Properties/MissingReadOnlyByPhpDocPropertyAssignRule.php:88]: Yes, I had a similar idea

**Comment by ondrejmirtes** [comment]
Thank you!

**Comment by ondrejmirtes** [comment]
Can you please send an update to this page @herndlm? https://phpstan.org/writing-php-code/phpdocs-basics

Thank you!

**Comment by herndlm** [comment]
I'll update the docs tomorrow, have fun in Verona!
